File: 916_word_subsets.java
class Solution {
    public List<String> wordSubsets(String[] w1, String[] w2) {
        int[] g = new int[26];
        for (String z : w2) {
            int[] c = map(z);
            for (int i = 0; i < 26; i++) g[i] = Math.max(g[i], c[i]);
        }
        List<String> out = new ArrayList<>();
        for (String q : w1) {
            int[] c = map(q);
            boolean ok = true;
            for (int i = 0; i < 26; i++) if (g[i] > c[i]) ok = false;
            if (ok) out.add(q);
        }
        return out;
    }
    private int[] map(String s) {
        int[] f = new int[26];
        for (char x : s.toCharArray()) f[x - 'a']++;
        return f;
    }
}

File: 916_word_subsets.cpp
class Solution {
public:
    array<int, 26> freqMap(string t) {
        array<int, 26> f = {0};
        for (auto e : t) f[e - 'a']++;
        return f;
    }
    vector<string> wordSubsets(vector<string>& a, vector<string>& b) {
        array<int, 26> limit = {0};
        for (auto& d : b) {
            auto g = freqMap(d);
            for (int i = 0; i < 26; i++) limit[i] = max(limit[i], g[i]);
        }
        vector<string> o;
        for (auto& x : a) {
            auto g = freqMap(x);
            bool ok = true;
            for (int i = 0; i < 26 && ok; i++) if (limit[i] > g[i]) ok = false;
            if (ok) o.push_back(x);
        }
        return o;
    }
};
